### PR TITLE
test: fix flaking e2e test (DRAFT)

### LINKED ISSUE #900: e2e test 'npt_to_port_22' is flaking
See #901 for updates on investigation and fix 

### Describe the bug

flaking

### Steps to reproduce

sometimes runs OK, sometimes does not

### Expected behavior

Should either fail for a valid reason, or consistently succeed

### Screenshots

_No response_

### Smartphones

_No response_

### Were you using an atApplication when the bug was found?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by gkc**
**- What I did**
- fix the flakiness of the npt_to_port_22 test
- Fixes #900 
- There is more to do in the future with regards to making diagnosis easier. The changes in this PR have made it much easier to diagnose problems which are srv-related, but we need to do more. I will noodle on this for a while.

**- How I did it**
- Actual fix is in [this commit](https://github.com/atsign-foundation/noports/pull/901/commits/2f4228f8a418c948db6c49713a1026ef5c0bd8a8)
- Everything else in this PR is enhancements of programmes' log output and of the test framework itself to make it possible to even diagnose what the actual problem was
  - Significant changes were made to the logging approach of packages/dart/sshnoports/bin/srv.dart - it now logs to a tmp file as well as (initially) to stderr; later on it stops logging to stderr and only to the tmp file. If it exits cleanly then it removes the tmp file; if not it leaves the tmp file in place so we have some hope of doing diagnosis.

**- How to verify it**
- Tests should not fail unless there is a genuine problem

### CODE REVIEW

**Comment by gkc** [comment]
Problem identified, cause not yet clear. Log excerpt from [this test run](https://github.com/atsign-foundation/noports/actions/runs/8419937665/job/23053963152)
```
2024-03-25 12:29:57.515 | Test 27 of 72 | Test script: npt_to_port_22 Client: Dart  (branch) Daemon: Dart  (branch)
2024-03-25 12:29:57.546 | Executing /home/ubuntu/noports/tests/e2e_all/runtime/29d8e7d7/binaries/branch/npt -f @apricoteventual -d 29d8e7d7dcf  -t @combinedbelgian -r @rv_am  --root-domain root.atsign.org --remote-port 22 --exit-when-connected
2024-03-25 12:30:01.571 | Executing ssh -p 35233 -o StrictHostKeyChecking=accept-new -o IdentitiesOnly=yes ubuntu@localhost -i /home/ubuntu/.ssh/e2e_all.29d8e7d7
    test execution's stdout: 
        2024-03-25 12:29:57.546 | Executing /home/ubuntu/noports/tests/e2e_all/runtime/29d8e7d7/binaries/branch/npt -f @apricoteventual -d 29d8e7d7dcf  -t @combinedbelgian -r @rv_am  --root-domain root.atsign.org --remote-port 22 --exit-when-connected

        2024-03-25 12:30:01.546 | Running ps for the spawned srv process BEFORE running ssh
        ==> NOTE THAT THERE IS NO srv process running for the port number which npt generated (35233)

        2024-03-25 12:30:01.571 | Executing ssh -p 35233 -o StrictHostKeyChecking=accept-new -o IdentitiesOnly=yes ubuntu@localhost -i /home/ubuntu/.ssh/e2e_all.29d8e7d7
        2024-03-25 12:30:01.602 | Running ps for the spawned srv process AFTER running ssh
    test execution's stderr: 
Connecting ... Connected
        2024-03-25 12:29:58.883168 : Sending daemon feature check request
        2024-03-25 12:29:58.885847 : Fetching host and port from srvd
        2024-03-25 12:30:00.123866 : Received host and port from srvd
        2024-03-25 12:30:00.123932 : Waiting for daemon feature check response
        2024-03-25 12:30:00.123941 : Received daemon feature check response
        2024-03-25 12:30:00.125941 : Required daemon features are supported
        2024-03-25 12:30:00.653588 : Sending session request to the device daemon
        2024-03-25 12:30:00.976026 : Waiting for response from the device daemon
        2024-03-25 12:30:01.429323 : Received response from the device daemon
        2024-03-25 12:30:01.429723 : Finding an available local port
        2024-03-25 12:30:01.429778 : Creating connection to socket rendezvous
        2024-03-25 12:30:01.533381 : requested localPort 0 ;  actual localPort 35233
        2024-03-25 12:30:01.533473 : Cleaning up temporary files

        ==> FAILING because the srv is not running! ssh: connect to host localhost port 35233: Connection refused

```

**Comment by gkc** [comment]
By wrapping the srv in a shell script and having the client execute that shell script instead of the srv binary, I have determined that there is indeed an unhandled exception somewhere (because the exit code is 255). What I don't understand yet is why the runZonedGuarded error handler does not appear to be handling that exception

For reference, the shell script
```
#!/bin/bash

# Sometimes while testing it's handy to execute this script rather than executing the srv directly

binDir="$(dirname -- "$0")"

mkdir -p /tmp/noports

# "${binDir}/srv" "$@" 2>&1 | tee -a /tmp/noports/srv.sh.$$.log
"${binDir}/srv" "$@"
echo "Exit code was $? " >> /tmp/noports/srv.sh.$$.log
```

**Comment by gkc** [comment]
I've finally got a smoking gun after making some substantial changes to how the logging works in the srv process.

The core problem is that the parent process (sshnp or npt) reads what it needs from the child process's stderr and then exits but somehow, sometimes, this happens _*before*_ the child process has finished using stderr.

```
SHOUT|2024-03-27 14:55:40.236002| srv.main |Unhandled exception FileSystemException: writeFrom failed, path = '' (OS Error: Broken pipe, errno = 32); stackTrace follows
#0      _RandomAccessFile.writeFromSync (dart:io/file_impl.dart:905)
#1      _StdConsumer.addStream.<anonymous closure> (dart:io/stdio.dart:309)
#2      _rootRunUnary (dart:async/zone.dart:1407)
#3      _CustomZone.runUnary (dart:async/zone.dart:1308)
#4      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1217)
#5      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:339)
#6      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271)
#7      _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:784)
#8      _StreamController._add (dart:async/stream_controller.dart:658)
#9      _StreamController.add (dart:async/stream_controller.dart:606)
#10     _StreamSinkImpl.add (dart:io/io_sink.dart:154)
#11     _IOSinkImpl.write (dart:io/io_sink.dart:287)
#12     _IOSinkImpl.writeln (dart:io/io_sink.dart:308)
#13     _StdSink.writeln (dart:io/stdio.dart:342)
#14     SrvImplDart.run (package:noports_core/src/srv/srv_impl.dart:421)
```

srv_impl.dart line 421 is : `stderr.writeln(Srv.startedString);`

File: packages/dart/noports_core/lib/src/srv/srv_impl.dart [hunk 1]
```diff
@@ -142,7 +142,9 @@ class SrvImplExec implements Srv<Process> {
       }
     });
 
-    await rvPortBound.future.timeout(Duration(seconds: 2));
+    await rvPortBound.future.timeout(Duration(seconds: 3));
+
+    await Future.delayed(Duration(milliseconds: 100));
```
Comment (gkc) [inline packages/dart/noports_core/lib/src/srv/srv_impl.dart:147]: A short grace period which should mean that the child process doesn't encounter an exception from stderr.writeln

File: packages/dart/noports_core/lib/src/srv/srv_impl.dart [hunk 2]
```diff
@@ -418,12 +420,19 @@ class SrvImplDart implements Srv<SocketConnector> {
...
       if (detached) {
-        stderr.writeln(Srv.startedString);
+        try {
+          stderr.writeln(Srv.startedString);
+        } catch (e, st) {
+          logger.severe('Failed to write ${Srv.startedString}'
+              ' to stderr: ${e.toString()} ;'
+              ' stackTrace follows:\n'
+              '$st');
+        }
```
Comment (gkc) [inline packages/dart/noports_core/lib/src/srv/srv_impl.dart:430]: If the child process _does_ encounter an exception from stderr.writeln, then catch it, log it, but continue to execute

**Review by xavierchanth — APPROVED** [review]

**Pull request merged**